### PR TITLE
caddytls: add 'key_type' subdirective

### PR DESCRIPTION
**Pull request by masipcat**
It seems that `key_type` subdirective it's not implemented in v2. I don't know if this is intentional but I needed this ASAP and I implemented it myself. Here is the code if you are interested

### CODE REVIEW

**Comment by mholt** [comment]
Thanks for the PR. Did you know [there's a global option to change `key_type`](https://caddyserver.com/docs/caddyfile/options#key-type)? Usually Caddyfile users just want to change the key type for all their sites; currently the JSON lets you configure key types per-domain/per-site, as most advanced users do.

What is your use case?

**Comment by masipcat** [comment]
Yes, I know about the global configuration and that it can be changed per-site using JSON but I prefer to continue with the Caddyfile for now :P 

My use case is: I have a VPS with a single IP address. I run multiple services (not only http server) and I use Caddy to get let's encrypt certificates for some of these other services. I want the EC certificates for the domains/subdomains served from caddy and the RSA for the other services.

File: caddyconfig/httpcaddyfile/builtins.go [hunk 1]
```diff
@@ -87,6 +87,7 @@ func parseTLS(h Helper) ([]ConfigValue, error) {
 	var folderLoader caddytls.FolderLoader
 	var certSelector caddytls.CustomCertSelectionPolicy
 	var acmeIssuer *caddytls.ACMEIssuer
+	var autoPolicy *caddytls.AutomationPolicy
```
Comment (mholt) [inline caddyconfig/httpcaddyfile/builtins.go:90]: The only field being used is the KeyType; Instead, we can probably just make a `keyType` variable here.

**Review by mholt — CHANGES_REQUESTED** [review]
Looks pretty good, but can simplify it slightly by not using an entire automation policy where we only need a key type variable.

**Comment by francislavoie** [comment]
Please also add a caddyfile adapt test, see here: https://github.com/caddyserver/caddy/tree/master/caddytest/integration/caddyfile_adapt

File: caddyconfig/httpcaddyfile/builtins.go [hunk 2]
```diff
@@ -87,7 +87,7 @@ func parseTLS(h Helper) ([]ConfigValue, error) {
 	var folderLoader caddytls.FolderLoader
 	var certSelector caddytls.CustomCertSelectionPolicy
 	var acmeIssuer *caddytls.ACMEIssuer
-	var autoPolicy *caddytls.AutomationPolicy
+	var keyType *string
```
Comment (mholt) [inline caddyconfig/httpcaddyfile/builtins.go:90]: Sorry to be pendantic, but why is this a pointer? :sweat_smile:  Probably safer to use a `string` instead. Just check for `keyType != ""`, impossible to panic accidentally.

```suggestion
	var keyType string
```

File: caddyconfig/httpcaddyfile/builtins.go [hunk 3]
```diff
@@ -395,10 +392,10 @@ func parseTLS(h Helper) ([]ConfigValue, error) {
 		})
 	}
 
-	if autoPolicy != nil {
+	if keyType != nil {
```
Comment (mholt) [inline caddyconfig/httpcaddyfile/builtins.go:395]:
```suggestion
	if keyType != "" {
```

File: caddyconfig/httpcaddyfile/builtins.go [hunk 4]
```diff
@@ -395,10 +392,10 @@ func parseTLS(h Helper) ([]ConfigValue, error) {
...
 		configVals = append(configVals, ConfigValue{
 			Class: "tls.key_type",
-			Value: autoPolicy.KeyType,
+			Value: *keyType,
```
Comment (mholt) [inline caddyconfig/httpcaddyfile/builtins.go:398]:
```suggestion
			Value: keyType,
```

File: caddyconfig/httpcaddyfile/builtins.go [hunk 5]
```diff
@@ -273,10 +273,7 @@ func parseTLS(h Helper) ([]ConfigValue, error) {
 				if len(arg) != 1 {
 					return nil, h.ArgErr()
 				}
-				if autoPolicy == nil {
-					autoPolicy = new(caddytls.AutomationPolicy)
-				}
-				autoPolicy.KeyType = arg[0]
+				*keyType = arg[0]
```
Comment (mholt) [inline caddyconfig/httpcaddyfile/builtins.go:276]:
```suggestion
				keyType = arg[0]
```

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks! Almost there. We can improve the safety of this code a bit.

File: caddyconfig/httpcaddyfile/builtins.go (hunk 2 above)
Comment (masipcat) [reply caddyconfig/httpcaddyfile/builtins.go:90]: Not a problem. I'm new to golang :)

**Review by mholt — APPROVED** [review]
Looks good now! Thanks very much. Not bad for a Go beginner. :)

**Comment by masipcat** [comment]
Thank you! :)

**Pull request merged**